src/dimension/broadcast.rs

```diff
@@ -0,0 +1,113 @@
+use crate::error::*;
+use crate::{Dimension, Ix0, Ix1, Ix2, Ix3, Ix4, Ix5, Ix6, IxDyn};
+
+/// Calculate the co_broadcast shape of two dimensions. Return error if shapes are
+/// not compatible.
+fn broadcast_shape<D1, D2, Output>(shape1: &D1, shape2: &D2) -> Result<Output, ShapeError>
+where
+    D1: Dimension,
+    D2: Dimension,
+    Output: Dimension,
+{
+    let (k, overflow) = shape1.ndim().overflowing_sub(shape2.ndim());
+    // Swap the order if d2 is longer.
+    if overflow {
+        return broadcast_shape::<D2, D1, Output>(shape2, shape1);
+    }
+    // The output should be the same length as shape1.
+    let mut out = Output::zeros(shape1.ndim());
+    let out_slice = out.slice_mut();
+    let s1 = shape1.slice();
+    let s2 = shape2.slice();
+    // Uses the [NumPy broadcasting rules]
+    // (https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html#general-broadcasting-rules).
+    //
+    // Zero dimension element is not in the original rules of broadcasting.
+    // We currently treat it like any other number greater than 1. As numpy does.
+    for i in 0..shape1.ndim() {
+        out_slice[i] = s1[i];
+    }
+    for i in 0..shape2.ndim() {
```

Comment:
index loop that should be a `zip(&mut out_slice[k..], &s2)` loop

Reply:
OK. Thanks a lot for taking the time to review.